### ISSUE TITLE
Problems encountered when freezing some parameters of the model during training

### ISSUE THREAD

**Issue by JWargrave** [huggingface/accelerate#1347]
I am new to `accelerate` and I've found that it really improves my development productivity. Thanks for your great work.

But I have some problems when freezing some parameters of the model during training.

Suppose some of the parameters are kept frozen while others are optimized during training. And I trained on four GPUs with `CUDA_VISIBLE_DEVICES="0,1,2,3" accelerate launch --multi_gpu --num_processes=4 main.py`.

At first, I forgot to call `model.train()` before the training loops.But the accuracy on the validation set is generally consistent with the results I got from training on a single GPU withouting using `accelerate`.

However, when I call `model.train()` before training loops, the accuracy on the validation set drops noticeably (I have tried for many times and the performance gap always exist).

The pseudocode of the training process is as follows:

```python
# model to train: model
accelerator=Accelerator(mixed_precision="fp16")

for some_param in model.parameters():
    some_param.requires_grad=False
for other_param in model.parameters():
    other_param.requires_grad=True

model.eval()
if accelerator.is_main_process:
    # extract features from validation set using model and these features will be used during the validation stage
    ...

model,...=accelerator.prepare(model,...)
    
for epoch in range(num_epochs):
    # ===== train =====
    model.train() # with or without **model.train()** makes a noticeable difference in accuracy on validation set
    for data,label in train_dataloader:
        with accelerator.autocast():
            loss=model(data,label)
        accelerator.backward(loss)
        optimizer.step()
    # ===== eval =====
    accelerator.wait_for_everyone()
    if accelerator.is_main_process:
        unwrapped_model=accelerator.unwrap_model(model)
        unwrapped_model.eval()
        # calculate accuracy on validation set using unwrapped_model:
        for data,label in val_dataloader:
            ...
    
```

The phenomenon above is very strange and I want to know the reason behind this and how to solve this problem.

Thanks a lot.

**Comment by muellerzr** [comment]
As noted in this doc, please be mindful of batch size and total batch size when comparing. https://huggingface.co/docs/accelerate/concept_guides/performance

Also, do note that fp16 won't get exactly the same training throughout, but the final result should be similar

**Comment by JWargrave** [comment]
@muellerzr Thank you for reply.
But the problem is that when I only change `model.train()` (add or remove), the performance changes noticeably.

**Comment by muellerzr** [comment]
How are you creating your optimizer?

**Comment by JWargrave** [comment]
@muellerzr 
```python
optimizer = optim.AdamW(
        [{'params': filter(lambda p: p.requires_grad, model.parameters()), 'lr': learning_rate,
          'betas': (0.9, 0.999), 'eps': 1e-7}])
```

**Comment by JWargrave** [comment]
@muellerzr @sgugger 
I have three questions: 
1. Whether all parameters of the model passed to `accelerator.prepare` should be trainable (i.e. requires_grad=True)?
2. If a model has no parameters with gradients, is passing it to `accelerator.prepare` or not exactly the same? (is [link](https://github.com/huggingface/accelerate/issues/1345#issuecomment-1518110375) true?)
3. Whether all parameters of the model passed to `accelerator.prepare` should be all trainable or all untrainable (i.e. have the same requires_grad)?

**Comment by JWargrave** [comment]
More findings:
```python
model.eval()
print(model.training) # False

model,...=accelerator.prepare(model,...)
print(model.training) # True

unwrapped_model=accelerator.unwrap_model(model)
print(unwrapped_model.training) # False

model.train()
print(model.training) # True
print(unwrapped_model.training) # True

unwrapped_model.eval()
print(model.training) # True
print(unwrapped_model.training) # False
```
Call `model.train()` or `model.eval()` will change both `model.training` and `unwrapped_model.training`.

But call `unwrapped_model.train()` or `unwrapped_model.eval()` will only change `unwrapped_model.training` and not change `model.training`.

Are these the expected results?

**Comment by sgugger** [comment]
Yes, that is completely normal.

**Comment by SangbumChoi** [comment]
@JWargrave Hi did you solve this problem?
